test/functional/api/map_spec.lua

```diff
@@ -0,0 +1,115 @@
+
+local helpers = require('test.functional.helpers')(after_each)
+local clear = helpers.clear
+local nvim = helpers.nvim
+local eq = helpers.eq
+local funcs = helpers.funcs
+
+local function local_copy(t)
+  local copy = {}
+  for k,v in pairs(t) do
+    copy[k] = v
+  end
+  return copy
+end
+
+describe('maparg and api', function()
+  before_each(clear)
+
+  local foo_bar_string = 'nnoremap foo bar'
+  local foo_bar_map_table = {
+      lhs='foo',
+      silent=0,
+      rhs='bar',
+      expr=0,
+      sid=0,
+      buffer=0,
+      nowait=0,
+      mode='n',
+      noremap=1,
+    }
+
+  -- change the map slightly and add a <nowait> option to the map
+  local foolong_barlong_string = 'nnoremap <nowait> foo_longer bar_longer'
+  local foolong_bar_map_table = local_copy(foo_bar_map_table)
+  foolong_bar_map_table['lhs'] = 'foo_longer'
+  foolong_bar_map_table['rhs'] = 'bar_longer'
+  foolong_bar_map_table['nowait'] = 1
+
+  local foo_bar_buffer_string = 'nnoremap <buffer> foo bar'
+  -- Only the buffer mapping should show up
+  local buffer_table = local_copy(foo_bar_map_table)
+  buffer_table['buffer'] = 1
+
+  -- The table will be the same except for the mode
+  local foo_bar_insert_string = 'inoremap foo bar'
+  local insert_table = local_copy(foo_bar_map_table)
+  insert_table['mode'] = 'i'
```

Comment:
And `<expr>`. And, especially, mapping with special characters like `<SID>` (need `helpers.source`) and `<F12>`.